### PR TITLE
KEYCLOAK-8181: Add spinner/loading indicator

### PR DESCRIPTION
**Pull request by ssilvert**

### CODE REVIEW

File: themes/src/main/resources/theme/keycloak-preview/account/messages/messages_en.properties [hunk 1]
```diff
@@ -9,6 +9,7 @@ done=Done
 cancel=Cancel
 remove=Remove
 update=Update
+loadingMessage=Keycloak Account Console is Loading ...
```
Comment (vmuzikar) [inline themes/src/main/resources/theme/keycloak-preview/account/messages/messages_en.properties]: I believe "loading" shouldn't start with capital L.

Comment (vmuzikar) [reply themes/src/main/resources/theme/keycloak-preview/account/messages/messages_en.properties]: Perhaps do we even need this message? One more thing to productize... IMHO just a spinner animation is clear enough.

**Review by vmuzikar — CHANGES_REQUESTED** [review]
Thanks, @ssilvert!

Unfortunately, before the spinner is displayed, 3.5 MB of data is downloaded, incl. e.g. 0.5 MB of PatternFly. Do we really need 3.5 MB just to display a spinner?

To understand what I mean and for steps to reproduce please see: https://youtu.be/AoE5u-NNs9o

**Comment by ssilvert** [comment]
> Thanks, @ssilvert!
> 
> Unfortunately, before the spinner is displayed, 3.5 MB of data is downloaded, incl. e.g. 0.5 MB of PatternFly. Do we really need 3.5 MB just to display a spinner?
> 
> To understand what I mean and for steps to reproduce please see: https://youtu.be/AoE5u-NNs9o

By default for now, it's in development mode.  You need to go into theme.properties and change developmentMode=false.  Then it will only download about 750K.  

I did a quick and dirty test on this to see how fast the spinner comes up.  This test has disabled browser caching and I hit refresh on a react page.  That's the slowest situation we could have:
dev mode slow 3G - 83 seconds
reg mode slow 3G - 38 seconds
reg mode fast 3G - 11 seconds

Of course, in the real world browser caching would be enabled so things would be faster than above.  Plus slow 3G is uncommon.

IMO, this is easily good enough for tech preview.  We will be revisiting performance issues anyway.

I think we should keep the text message.  It's comforting.  As for the capital 'L', I can go either way.

**Comment by vmuzikar** [comment]
@ssilvert +1 that's more than sufficient for a tech preview.

Unfortunately, I believe slow mobile internet speed is still common at many place (heck, even I'm currently sitting in a blank space of the map where all you can get is 2G :D). So I'd like loading UXD (which is the first thing user encounters) to be as best as possible in given environment.
Would it be technically possible to first load the bare minimum required for displaying just the spinner? If so, we should create a follow-up JIRA.

**Comment by ssilvert** [comment]
> @ssilvert +1 that's more than sufficient for a tech preview.
> 
> Unfortunately, I believe slow mobile internet speed is still common at many place (heck, even I'm currently sitting in a blank space of the map where all you can get is 2G :D). So I'd like loading UXD (which is the first thing user encounters) to be as best as possible in given environment.
> Would it be technically possible to first load the bare minimum required for displaying just the spinner? If so, we should create a follow-up JIRA.

I agree.  We will need to follow up on this.  If needed, I could reimplement this without patternfly.min.cs.  It looks like loading that file is the current bottleneck.

**Review by vmuzikar — DISMISSED** [review]
Created https://issues.redhat.com/browse/KEYCLOAK-12476

Thanks @ssilvert!

**Comment by ssilvert** [comment]
@vmuzikar I went ahead and implemented the performance enhancement over the weekend.  It now:

- Delays loading patternfly-min.css so that you can see the spinner sooner.  
- Turns off development mode by default.
- The spinner show up any time there is a transition, including when the hello screen is loading.

This should be much better now for 3G.  Let me know if you think it's good enough to close https://issues.redhat.com/browse/KEYCLOAK-12476.

BTW, "Slow 3G" is extremely slow.  With the old console, it takes about 16 seconds to see anything.  

With the new console, it now takes only 10 seconds to see the spinner.  But it still takes way too long to load the full app.

On "Fast 3G", it's 2.5 seconds to see the spinner and 40 seconds to load the app.

We will still need to work on performance, but at least the spinner logic is improved.

**Review by vmuzikar — APPROVED** [review]
@ssilvert Thanks, it indeed works much better now. I've closed [KEYCLOAK-12476](https://issues.redhat.com/browse/KEYCLOAK-12476).

I've noticed one last small issue. When the welcome screen is loaded (i.e. the spinner is no longer displayed) but the React app is still loading and the user clicks a link on the welcome page, seemingly nothing happens (because the React app is not yet loaded and spinner is gone by this time). Would it be possible to display the spinner again to explicitly show the user that something is happening? This is rather an enhancement and could be done as a follow-up task, so I'm gonna approve beforehand (I'm on PTO from tomorrow). In case you decided to do this later, please create a JIRA so we won't forget about this.

**Comment by ssilvert** [comment]
>
> I've noticed one last small issue. When the welcome screen is loaded (i.e. the spinner is no longer displayed) but the React app is still loading and the user clicks a link on the welcome page, seemingly nothing happens (because the React app is not yet loaded and spinner is gone by this time). 

I actually fixed that problem with this latest update.  I call toggleReact() when the link is clicked and it re-displays the spinner.

If you switch to fast 3G you will see that it does display the spinner again before React is loaded.  It looks like the problem is that on slow 3G it's unable to repaint the screen.  Not sure what to do about that.

**Review by abstractj — APPROVED** [review]

**Pull request merged**